### PR TITLE
test: added test for upgradewallet RPC

### LINKED ISSUE #18767: test: Write tests for upgradewallet RPC
Now that #15761 has been merged, it would be nice to have tests for it. They can be written by copying the file `./test/functional/feature_backwards_compatibility.py` and modifying it to create a wallet with an older version, then check that it can be upgraded with a new version.

#### Useful skills:

Experience with the Bitcoin Core wallet and functional test suite.

#### Want to work on this issue?

The purpose of the `good first issue` label is to highlight which issues are suitable for a new contributor without a deep understanding of the codebase.

You do not need to request permission to start working on this. You are encouraged to comment on the issue if you are planning to work on it. This will help other contributors monitor which issues are actively being addressed and is also an effective way to request assistance if and when you need it.

For guidance on contributing, please read [CONTRIBUTING.md](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md) before opening your pull request.

### PR DESCRIPTION
**Pull request by brakmic**
This PR adds tests for the newly merged *upgradewallet* RPC. 

Additionally, it expands `test_framework/util.py` by adding the function `adjust_bitcoin_conf_for_pre_17` to support nodes that don't parse configuration sections. 

This test uses two older node versions, v0.15.2 and v0.16.3, to create older wallet versions to be used by `upgradewallet`.

Fixes https://github.com/bitcoin/bitcoin/issues/18767

### CODE REVIEW

File: test/functional/wallet_upgradewallet.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            160300
+        ], binary=[
+            self.options.bitcoind,
+            releases_path + "/v0.16.3/bin/bitcoind",
+        ], binary_cli=[
+            self.options.bitcoincli,
+            releases_path + "/v0.16.3/bin/bitcoin-cli",
+        ])
+
+        self.start_nodes()
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:54]: Instead of modifying the test_framwork/util.py, you could just remove the line `[regtest]` from the bitcoin.conf in the 16.3 datadir before starting the nodes

Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:54]: Yes, but wouldn't be better to make this generally available. For example the [comment](https://github.com/bitcoin/bitcoin/blob/master/test/functional/feature_backwards_compatibility.py#L11) in feature_backwards_compatibility states that it's not possible to test older nodes.

Comment (MarcoFalke) [reply test/functional/wallet_upgradewallet.py:56]: Yes, if you want, you can create a generally available util method like

```py
def adjust_bitcoin_conf_for_pre_16_3(conf_file):
  ...

File: test/functional/wallet_upgradewallet.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+            160300
+        ], binary=[
+            self.options.bitcoind,
+            releases_path + "/v0.16.3/bin/bitcoind",
+        ], binary_cli=[
+            self.options.bitcoincli,
+            releases_path + "/v0.16.3/bin/bitcoin-cli",
+        ])
+        # older bitcoind's like v0.16.3 don't recognize config sections
+        subprocess.call(["sed","-i", "-e", "s/\[regtest\]/ /g", self.nodes[1].bitcoinconf])
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:55]: sed does not exist on windows. I think this is possible entirely in python?

File: test/functional/wallet_upgradewallet.py (hunk 1 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:56]: Ok, will implement it. Have just force pushed a change that uses *sed* to make an in-file replacement.

File: test/functional/wallet_upgradewallet.py (hunk 2 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:55]: Yes, just came to my mind. Windows!
I'll use python's own regex utils.

**Comment by brakmic** [comment]
@MarcoFalke 
There are several error messages in travis that I don't know how to fix. Maybe because the v0.16.3 binaries are missing on test envs?

File: test/functional/test_runner.py [hunk 3]
```diff
@@ -180,6 +180,7 @@
     'mempool_expiry.py',
     'wallet_import_rescan.py',
     'wallet_import_with_label.py',
+    'wallet_upgradewallet',
```
Comment (MarcoFalke) [inline test/functional/test_runner.py:183]:
```suggestion
    'wallet_upgradewallet.py',
```

**Comment by MarcoFalke** [comment]
You'll also need to adjust ci/test/05_before_script.sh

File: test/functional/test_framework/util.py [hunk 4]
```diff
@@ -326,6 +326,16 @@ def initialize_datadir(dirname, n, chain):
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
 
+def adjust_bitcoin_conf_for_pre_16_3(conf_file):
+    conf = open(conf_file,'r', encoding='utf8')
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:330]: python `with` context manager should simplify this

File: test/functional/wallet_upgradewallet.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+    def setup_nodes(self):
+        if os.getenv("TEST_PREVIOUS_RELEASES") == "false":
+            raise SkipTest("upgradewallet RPC tests")
+
+        releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
+        if not os.path.isdir(releases_path):
+            if os.getenv("TEST_PREVIOUS_RELEASES") == "true":
+                raise AssertionError("TEST_PREVIOUS_RELEASES=1 but releases missing: " + releases_path)
+            raise SkipTest("This test requires binaries for previous releases")
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:42]: Can this be factored out into a `def skip_if_no_previous_release(self)`? Maybe a future pull?

File: test/functional/wallet_upgradewallet.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+        self.log.info("Test upgradewallet RPC...")
+        # Prepare for copying of the older wallet
+        node_master_wallets_dir = os.path.join(node_master.datadir, "regtest/wallets")
+        v16_3_wallets_dir       = os.path.join(v16_3_node.datadir, "regtest/wallets")
+        node_master.unloadwallet("")
+
+        # Copy the 0.16.3 wallet to the last Bitcoin Core version and open it:
+        shutil.rmtree(node_master_wallets_dir)
+        shutil.copytree(
+            v16_3_wallets_dir,
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:75]: don't you need to shut down first?

File: test/functional/wallet_upgradewallet.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+        releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
+        if not os.path.isdir(releases_path):
+            if os.getenv("TEST_PREVIOUS_RELEASES") == "true":
+                raise AssertionError("TEST_PREVIOUS_RELEASES=1 but releases missing: " + releases_path)
+            raise SkipTest("This test requires binaries for previous releases")
+
+        self.add_nodes(self.num_nodes, extra_args=self.extra_args, versions=[
+            None,
+            160300
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:46]: would it be possible to user version 15, so that `-usehd=0` can be passed to the wallet?

File: test/functional/test_framework/util.py (hunk 4 above)
Comment (brakmic) [reply test/functional/test_framework/util.py:330]: Have changed it.

File: test/functional/wallet_upgradewallet.py (hunk 5 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:44]: I've taken this piece of code from `feature_backwards_compatibility`, so I will later try it out with `skip_if_no_previous_release(self)`.

File: test/functional/wallet_upgradewallet.py (hunk 6 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:86]: Indeed. The latest variant has it now.

File: test/functional/wallet_upgradewallet.py (hunk 7 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py]: Have added a v0.15.2 node, so that we are now testing v0.15.2 with *-usehd=0* and v0.16.3 with *-usehd=1*.

Comment (brakmic) [reply test/functional/wallet_upgradewallet.py]: Btw. there is now a new helper method in *util.py* called `get_nodes_least_version`, because the framework couldn't start the v0.15.2 node. The error was thrown in `sync_mempools`and `sync_blocks`. It seems to have something to do with the call of `syncwithvalidationinterfacequeue`. This indirectly affected the `getpeerinfo`from `sync_blocks`, but because 0.15.2 never needed any peers for these particular tests, I thought, maybe a small helper methode for filtering those old nodes would be sufficient. If not, please, let me know what alternatives we have here.

File: test/functional/test_framework/util.py [hunk 8]
```diff
@@ -326,6 +326,14 @@ def initialize_datadir(dirname, n, chain):
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
 
+def adjust_bitcoin_conf_for_pre_16_3(conf_file):
+    conf_data = ""
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:330]: This line can be removed, because python's scopes are more fluid

File: test/functional/test_framework/util.py [hunk 9]
```diff
@@ -326,6 +326,14 @@ def initialize_datadir(dirname, n, chain):
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
 
+def adjust_bitcoin_conf_for_pre_16_3(conf_file):
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:329]:
```suggestion
def adjust_bitcoin_conf_for_pre_17(conf_file):
```

Sorry, this should say pre 17

File: test/functional/test_framework/util.py [hunk 10]
```diff
@@ -408,13 +416,22 @@ def sync_blocks(rpc_connections, *, wait=1, timeout=60):
         if best_hash.count(best_hash[0]) == len(rpc_connections):
             return
         # Check that each peer has at least one connection
-        assert (all([len(x.getpeerinfo()) for x in rpc_connections]))
+        assert (all([len(x.getpeerinfo()) for x in get_nodes_least_version(rpc_connections)]))
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:419]: I don't like making sync_**all** only sync the latest versions. Why is this needed here anyway?

File: test/functional/wallet_upgradewallet.py (hunk 5 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:43]: I think, I should then also update every test where the above code is currently in use. Will make a separate pull request for that. :)

File: test/functional/test_framework/util.py [hunk 11]
```diff
@@ -426,8 +443,8 @@ def sync_mempools(rpc_connections, *, wait=1, timeout=60, flush_scheduler=True):
         pool = [set(r.getrawmempool()) for r in rpc_connections]
         if pool.count(pool[0]) == len(rpc_connections):
             if flush_scheduler:
-                for r in rpc_connections:
-                    r.syncwithvalidationinterfacequeue()
+                for r in get_nodes_least_version(rpc_connections):
+                        r.syncwithvalidationinterfacequeue()
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:447]: I'd rather remove the call to sync_mempools. You can do that by deriving the function that calls it.

**Review by maflcko — APPROVED** [review]

File: test/functional/test_framework/util.py (hunk 10 above)
Comment (brakmic) [reply test/functional/test_framework/util.py:419]: Because the v0.15.2 node has no peers. But you're right, if I am not using *sync_blocks* it should not react at all. In a previous version of upgradewallet test there was a *sync_blocks* call. Will adapt the code.

File: test/functional/test_framework/util.py (hunk 11 above)
Comment (brakmic) [reply test/functional/test_framework/util.py:447]: If I only knew, where am I calling *sync_mempools*. Will have to search a bit.

Comment (brakmic) [reply test/functional/test_framework/util.py:447]: I think, it comes indirectly from *setup_network*. This method also calls *sync_* stuff.

Comment (MarcoFalke) [reply test/functional/test_framework/util.py:447]: You can override setup_network and provide your own impl

File: test/functional/wallet_upgradewallet.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+        self.log.info("Test upgradewallet RPC...")
+        # Prepare for copying of the older wallet
+        node_master_wallets_dir = os.path.join(node_master.datadir, "regtest/wallets")
+        v16_3_wallets_dir       = os.path.join(v16_3_node.datadir, "regtest/wallets")
+        v15_2_wallets_dir       = os.path.join(v15_2_node.datadir, "regtest")
+        # stop all nodes
+        self.stop_node(0)
+        self.stop_node(1)
+        self.stop_node(2)
```
Comment (achow101) [inline test/functional/wallet_upgradewallet.py:83]: nit: `self.stop_nodes` instead of listing each individually?

File: test/functional/wallet_upgradewallet.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+
+        # Copy the 0.16.3 wallet to the last Bitcoin Core version and open it:
+        shutil.rmtree(node_master_wallets_dir)
+        shutil.copytree(
+            v16_3_wallets_dir,
+            node_master_wallets_dir
+        )
+        self.restart_node(0, ['-nowallet'])
+
+        node_master.loadwallet(node_master_wallets_dir)
```
Comment (achow101) [inline test/functional/wallet_upgradewallet.py:93]: Since you're just loading the default wallet, `loadwallet('')` should be sufficient.

File: test/functional/wallet_upgradewallet.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        assert_equal(res['blocks'], 101)
+        node_master = self.nodes[0]
+        v16_3_node  = self.nodes[1]
+        v15_2_node  = self.nodes[2]
+
+        self.log.info("Test upgradewallet RPC...")
+        # Prepare for copying of the older wallet
+        node_master_wallets_dir = os.path.join(node_master.datadir, "regtest/wallets")
+        v16_3_wallets_dir       = os.path.join(v16_3_node.datadir, "regtest/wallets")
+        v15_2_wallets_dir       = os.path.join(v15_2_node.datadir, "regtest")
```
Comment (achow101) [inline test/functional/wallet_upgradewallet.py:79]: This is the entire datadir and its going to be copying a lot more than just the wallet. Instead of copying the entire walletdir, just give the paths to the specific wallet.dat files?

**Review by achow101 — COMMENTED** [review]
We should be be testing the specific version upgrades too, but this is a good place to start with. Those can be added in later.

ACK 84267045efbfbe2a035e7f34851f1b3bc9964be8

File: test/functional/test_framework/util.py [hunk 15]
```diff
@@ -415,7 +422,6 @@ def sync_blocks(rpc_connections, *, wait=1, timeout=60):
         "".join("\n  {!r}".format(b) for b in best_hash),
     ))
 
-
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:418]: nit: why remove this line? Seems unrelated and goes against the pep8 rules

File: test/functional/test_framework/util.py [hunk 16]
```diff
@@ -326,6 +326,13 @@ def initialize_datadir(dirname, n, chain):
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
 
+def adjust_bitcoin_conf_for_pre_17(conf_file):
+    with open(conf_file,'r', encoding='utf8') as conf:
+        conf_data = conf.read()
+    with open(conf_file, 'w+', encoding='utf8') as conf:
+        conf_data_changed = re.sub(r'\[regtest\]', ' ', conf_data)
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:333]:
```suggestion
        conf_data_changed = conf_data.replace('[regtest]', '')
```

Should be doable without relying on the regex module

File: test/functional/wallet_upgradewallet.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+    def run_test(self):
+        self.nodes[0].generatetoaddress(101, self.nodes[0].getnewaddress())
+        # # Sanity check the test framework:
+        res = self.nodes[0].getblockchaininfo()
+        assert_equal(res['blocks'], 101)
+        node_master = self.nodes[0]
+        v16_3_node  = self.nodes[1]
+        v15_2_node  = self.nodes[2]
+
+        self.log.info("Test upgradewallet RPC...")
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:75]: It would be nice to send some coins around, so that we can check the balance doesn't decrease from upgrading the wallet

File: test/functional/wallet_upgradewallet.py [hunk 18]
```diff
@@ -0,0 +1,125 @@
...
+
+        node_master.loadwallet(node_master_wallets_dir)
+        wallet_name = node_master.getwalletinfo()["walletname"]
+        wallet = node_master.get_wallet_rpc(wallet_name)
+        # calling upgradewallet with explicit version number
+        # should return nothing if successful
+        assert_equal(wallet.upgradewallet(169900), "")
+        new_version = wallet.getwalletinfo()["walletversion"]
+        # upgraded wallet should have version 169900
+        assert_equal(new_version, 169900)
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:122]: Is this now an HD wallet? If yes, would be nice to check that the master key hash is present in the wallet info (or something like that)

**Review by maflcko — APPROVED** [review]
ACK

some more suggestions

File: test/functional/wallet_upgradewallet.py (hunk 17 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:75]: Sadly, this can't be implemented, because old nodes aren't connected to *node_master*. We had this problem with rpc_connections previously when I tried to filter out old nodes. Now, the same problem happens when trying to use *sync_* methods.

File: test/functional/wallet_upgradewallet.py (hunk 18 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:122]: Have added appropriate tests.

File: test/functional/wallet_upgradewallet.py (hunk 17 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:75]: More precisely, this is the place where the error gets thrown.

```shell
 File "/test/functional/test_framework/test_framework.py", line 499, in sync_blocks
    sync_blocks(nodes or self.nodes, **kwargs)
  File "/test/functional/test_framework/util.py", line 418, in sync_blocks
    assert (all([len(x.getpeerinfo()) for x in rpc_connections]))
```

We have seen this already. So, the question is, do we want to work on "onboarding" of old nodes so their converted wallets can be tested here, or should we rather open a dedicated PR for this?

File: test/functional/wallet_upgradewallet.py (hunk 14 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:79]: Yes, that's much better. Have adapted the code.

File: test/functional/wallet_upgradewallet.py (hunk 17 above)
Comment (MarcoFalke) [reply test/functional/wallet_upgradewallet.py:75]: So syncing the blockchain is the issue?

Can be done in a really dumb way:

```
def dumb_sync_blocks(...):
 node_from = select node with max height
 for i in range(max(blockheight)):
  b = node_from.getblock(lbockhash=node_from.getblockhash(i), verbose=0)
  for n in nodes:
    n.submitblock(b)

Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:75]: Thanks. Have tried it, but the v0.15.2 node doesn't expand its chain with submitted blocks. Interestingly, in its blockchaininfo the header count gets increased while block count remains at 0. The other node, v0.16.3 is working fine. Here my variant of *dumb_sync_blocks*

```python
def dumb_sync_blocks(self):
    node_from = self.nodes[0]
    to_height = node_from.getblockheader(node_from.getbestblockhash())['height']
    for n in range(1, len(self.nodes)):
        height = self.nodes[n].getblockheader(self.nodes[n].getbestblockhash())['height']
        for i in range(height, to_height+1):
            b = node_from.getblock(blockhash=node_from.getblockhash(i), verbose=0)
            self.nodes[n].submitblock(b)
        new_height = self.nodes[n].getblockheader(self.nodes[n].getbestblockhash())['height']
        print("node: {}, new best height: {}".format(n, new_height))
```

The output is like this:

```shell
node: 1, new best height: 101
node: 2, new best height: 0
```

Here their respective blockchaininfo jsons:

**v0.16.3**
```js
{
'chain': 'regtest', 
'blocks': 101, 
'headers': 101, 
'bestblockhash': '1a2f4021b111f20b12071bb68c92dc0037a2237ae034b345b002971dbb9f5e16', }
```

**v0.15.2**
```js
{
'chain': 'regtest', 
'blocks': 0, 
'headers': 101, 
'bestblockhash': '0f9188f13cb7b2c71f2a335e3a4fc328bf5beb436012afca590b1a11466e2206' }
```

In the test code I am sending some coins to both wallets and then calling *dumb_sync_blocks* to update all chains.

```python
node_master.generatetoaddress(101, v16_3_address)
self.dumb_sync_blocks()
```

Comment (MarcoFalke) [reply test/functional/wallet_upgradewallet.py:75]: Ah, right. regtest was hardforked

Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:75]: Ok, then I will be doing balance tests with v0.16.3 only? v0.15.2 can still be used for upgrade tests. Is this acceptable?

Comment (MarcoFalke) [reply test/functional/wallet_upgradewallet.py:75]: It might be possible to apply a hack and either:

* never produce witnesses (use legacy-1-addresses)
* maybe strip the witness from the blocks before submitting, but I doubt that works because VERIFY_WITNESS should be enforced from genesis, always.

Comment (MarcoFalke) [reply test/functional/wallet_upgradewallet.py:75]: You don't have to do this, I am happy to merge the pull in its current state. Just let me know.

Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:75]: Thanks. I think I should try to implement it. If it works, fine. If not, then you can merge the current state. Will come back to you later.

Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:75]: Well, I have managed to sync the chains, but there is no way to make proper transactions. I have stolen a few things from *p2p_segwit*, like *build_next_block* for manual block generation and syncing:

```js
node: 1, new best height: 101
node: 2, new best height: 101
{
'chain': 'regtest', 
'blocks': 101, 
'headers': 101, 
'bestblockhash': '0f2cfae3d63464c070eea6a8305f6f02245e69f41987c186da1c7f8af2bb1547', }
{
'chain': 'regtest', 
'blocks': 101, 
'headers': 101, 
'bestblockhash': '0f2cfae3d63464c070eea6a8305f6f02245e69f41987c186da1c7f8af2bb1547'}
``` 

However, I see no way how to update balance of v0.15.2 as I'd need to build a transaction that would have to have the witness field.

I will now push the latest changes.

File: test/functional/test_framework/util.py [hunk 19]
```diff
@@ -326,6 +326,13 @@ def initialize_datadir(dirname, n, chain):
         os.makedirs(os.path.join(datadir, 'stdout'), exist_ok=True)
     return datadir
 
+def adjust_bitcoin_conf_for_pre_17(conf_file):
+    with open(conf_file,'r', encoding='utf8') as conf:
+        conf_data = conf.read()
+    with open(conf_file, 'w+', encoding='utf8') as conf:
```
Comment (MarcoFalke) [inline test/functional/test_framework/util.py:332]:
```suggestion
    with open(conf_file, 'w', encoding='utf8') as conf:
```

no need for the read flag

File: test/functional/wallet_upgradewallet.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+        """
+        Little helper to sync older wallets. 
+        Notice that v0.15.2's regtest is hardforked, so there is 
+        actually no sync for it. v0.15.2 is only being used
+        to test version upgrade and master hash key presence.
+        v0.16.3 is being used to test version upgrade and balances.
+        Further info: https://github.com/bitcoin/bitcoin/pull/18774#discussion_r416967844
+        """
+        node_from = self.nodes[0]
+        to_height = node_from.getblockheader(node_from.getbestblockhash())['height']
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:77]:
```suggestion
        to_height = node_from.getblockcount()
```

File: test/functional/wallet_upgradewallet.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+        for n in range(1, len(self.nodes)):
+            height = self.nodes[n].getblockheader(self.nodes[n].getbestblockhash())['height']
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:79]:
```suggestion
            height = self.nodes[n].getblockcount()
```

File: test/functional/wallet_upgradewallet.py [hunk 22]
```diff
@@ -0,0 +1,155 @@
...
+            for i in range(height, to_height+1):
+                b = node_from.getblock(blockhash=node_from.getblockhash(i), verbose=0)
+                self.nodes[n].submitblock(b)
+
```
Comment (MarcoFalke) [inline test/functional/wallet_upgradewallet.py:83]:
```suggestion
            assert_equal(self.nodes[n].getblockcount(), to_height)
```

**Review by maflcko — APPROVED** [review]

File: test/functional/wallet_upgradewallet.py (hunk 22 above)
Comment (brakmic) [reply test/functional/wallet_upgradewallet.py:85]: Thanks. Btw., have tried to add you as Co-Author, but couldn't find your email or "no-reply" email.

**Review by maflcko — APPROVED** [review]
ACK. Will be merged as soon as travis is green
